File: CS/SchedulerInplaceEditorExample/MyInplaceEditorControl.cs
using System;
using System.Drawing;
using System.Windows.Forms;
using DevExpress.XtraScheduler;

namespace SchedulerInplaceEditorExample {
    #region #MyInplaceEditorControl
    public class MyInplaceEditorControl : ISchedulerInplaceEditorEx {
        MyInplaceEditor editor;
        Appointment appointment;
        SchedulerControl control;

        public MyInplaceEditorControl(SchedulerInplaceEditorEventArgs inplaceEditorArgs) {
            this.appointment = inplaceEditorArgs.ViewInfo.Appointment;
            this.control = inplaceEditorArgs.Control;
            CreateEditor(inplaceEditorArgs);
        }

        MyInplaceEditor Editor { get { return editor; } }
        Appointment Appointment { get { return appointment; } }
        SchedulerControl Control { get { return control; } }

        public event EventHandler CommitChanges;
        public event EventHandler RollbackChanges;

        void CreateEditor(SchedulerInplaceEditorEventArgs inplaceEditorArgs) {
            this.editor = new MyInplaceEditor(inplaceEditorArgs);
        }

        // Interface implementation 
        public void Dispose() {
            Dispose(true);
            GC.SuppressFinalize(this);
        }
        ~MyInplaceEditorControl() {
            Dispose(false);
        }
        protected virtual void Dispose(bool disposing) {
            if (disposing) {
                if (Editor != null) {
                    Editor.Dispose();
                    this.editor = null;
                }
                this.appointment = null;
            }
        }
        public virtual void Activate() {
            Editor.FillForm(control, appointment);
            SubscribeEditorEvents();
            Editor.Show(Control.FindForm());
        }
        public virtual void Deactivate() {
            UnsibscribeEditorEvents();
            Editor.Close();
        }
        public virtual void ApplyChanges() {
            Editor.ApplyChanges();
        }
        protected internal virtual void SubscribeEditorEvents() {
            Editor.FormClosed += new FormClosedEventHandler(Editor_FormClosed);
            Editor.Deactivate += new EventHandler(Editor_Deactivate);
            Editor.CommitChanges += new EventHandler(Editor_CommitChanges);
            Editor.RollbackChanges += new EventHandler(Editor_RollbackChanges);
        }

        protected internal virtual void UnsibscribeEditorEvents() {
            Editor.FormClosed -= new FormClosedEventHandler(Editor_FormClosed);
            Editor.Deactivate -= new EventHandler(Editor_Deactivate);
            Editor.CommitChanges -= new EventHandler(Editor_CommitChanges);
            Editor.RollbackChanges -= new EventHandler(Editor_RollbackChanges);
        }
        void Editor_FormClosed(object sender, FormClosedEventArgs e) {
            OnCommitChanges();
        }
        void Editor_Deactivate(object sender, EventArgs e) {
            OnCommitChanges();
        }
        void Editor_RollbackChanges(object sender, EventArgs e) {
            OnRollbackChanges();
        }
        void Editor_CommitChanges(object sender, EventArgs e) {
            OnCommitChanges();
        }
        protected internal virtual void TextBox_LostFocus(object sender, EventArgs e) {
            Editor.Close();
            OnCommitChanges();
        }
        protected internal virtual void OnRollbackChanges() {
            if (RollbackChanges != null)
                RollbackChanges(this, EventArgs.Empty);
        }
        protected internal virtual void OnCommitChanges() {
            RaiseCommitChanges();
        }
        protected internal virtual void RaiseCommitChanges() {
            if (CommitChanges != null)
                CommitChanges(this, EventArgs.Empty);
        }
    }
    #endregion #MyInplaceEditorControl
}

File: VB/SchedulerInplaceEditorExample/MyInplaceEditorControl.vb
﻿Imports System
Imports System.Drawing
Imports System.Windows.Forms
Imports DevExpress.XtraScheduler

Namespace SchedulerInplaceEditorExample
    #Region "#MyInplaceEditorControl"
    Public Class MyInplaceEditorControl
        Implements ISchedulerInplaceEditorEx


        Private editor_Renamed As MyInplaceEditor

        Private appointment_Renamed As Appointment

        Private control_Renamed As SchedulerControl

        Public Sub New(ByVal inplaceEditorArgs As SchedulerInplaceEditorEventArgs)
            Me.appointment_Renamed = inplaceEditorArgs.ViewInfo.Appointment
            Me.control_Renamed = inplaceEditorArgs.Control
            CreateEditor(inplaceEditorArgs)
        End Sub

        Private ReadOnly Property Editor() As MyInplaceEditor
            Get
                Return editor_Renamed
            End Get
        End Property
        Private ReadOnly Property Appointment() As Appointment
            Get
                Return appointment_Renamed
            End Get
        End Property
        Private ReadOnly Property Control() As SchedulerControl
            Get
                Return control_Renamed
            End Get
        End Property

        Public Event CommitChanges As EventHandler Implements ISchedulerInplaceEditorEx.CommitChanges
        Public Event RollbackChanges As EventHandler Implements ISchedulerInplaceEditorEx.RollbackChanges

        Private Sub CreateEditor(ByVal inplaceEditorArgs As SchedulerInplaceEditorEventArgs)
            Me.editor_Renamed = New MyInplaceEditor(inplaceEditorArgs)
        End Sub

        ' Interface implementation 
        Public Sub Dispose() Implements ISchedulerInplaceEditorEx.Dispose
            Dispose(True)
            GC.SuppressFinalize(Me)
        End Sub
        Protected Overrides Sub Finalize()
            Dispose(False)
        End Sub
        Protected Overridable Sub Dispose(ByVal disposing As Boolean)
            If disposing Then
                If Editor IsNot Nothing Then
                    Editor.Dispose()
                    Me.editor_Renamed = Nothing
                End If
                Me.appointment_Renamed = Nothing
            End If
        End Sub
        Public Overridable Sub Activate() Implements ISchedulerInplaceEditorEx.Activate
            Editor.FillForm(control_Renamed, appointment_Renamed)
            SubscribeEditorEvents()
            Editor.Show(Me.Control.FindForm())
        End Sub
        Public Overridable Sub Deactivate() Implements ISchedulerInplaceEditorEx.Deactivate
            UnsibscribeEditorEvents()
            Editor.Close()
        End Sub
        Public Overridable Sub ApplyChanges() Implements ISchedulerInplaceEditorEx.ApplyChanges
            Editor.ApplyChanges()
        End Sub
        Protected Friend Overridable Sub SubscribeEditorEvents()
            AddHandler Editor.FormClosed, AddressOf Editor_FormClosed
            AddHandler Editor.Deactivate, AddressOf Editor_Deactivate
            AddHandler Editor.CommitChanges, AddressOf Editor_CommitChanges
            AddHandler Editor.RollbackChanges, AddressOf Editor_RollbackChanges
        End Sub

        Protected Friend Overridable Sub UnsibscribeEditorEvents()
            RemoveHandler Editor.FormClosed, AddressOf Editor_FormClosed
            RemoveHandler Editor.Deactivate, AddressOf Editor_Deactivate
            RemoveHandler Editor.CommitChanges, AddressOf Editor_CommitChanges
            RemoveHandler Editor.RollbackChanges, AddressOf Editor_RollbackChanges
        End Sub
        Private Sub Editor_FormClosed(ByVal sender As Object, ByVal e As FormClosedEventArgs)
            OnCommitChanges()
        End Sub
        Private Sub Editor_Deactivate(ByVal sender As Object, ByVal e As EventArgs)
            OnCommitChanges()
        End Sub
        Private Sub Editor_RollbackChanges(ByVal sender As Object, ByVal e As EventArgs)
            OnRollbackChanges()
        End Sub
        Private Sub Editor_CommitChanges(ByVal sender As Object, ByVal e As EventArgs)
            OnCommitChanges()
        End Sub
        Protected Friend Overridable Sub TextBox_LostFocus(ByVal sender As Object, ByVal e As EventArgs)
            Editor.Close()
            OnCommitChanges()
        End Sub
        Protected Friend Overridable Sub OnRollbackChanges()
            RaiseEvent RollbackChanges(Me, EventArgs.Empty)
        End Sub
        Protected Friend Overridable Sub OnCommitChanges()
            RaiseCommitChanges()
        End Sub
        Protected Friend Overridable Sub RaiseCommitChanges()
            RaiseEvent CommitChanges(Me, EventArgs.Empty)
        End Sub
    End Class
    #End Region ' #MyInplaceEditorControl
End Namespace
